olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,244 @@
+            if model_path is None and model_config_copy.config.get("hf_config") is not None:
+                # hf model, get model commit from hf cache
+                cache_dir = Path(self.hf_cache_path)
+                model_cache_dir = Path(cache_dir) / "hub"
+                model_name = f"models--{model_config_copy.config['hf_config']['model_name'].replace('/', '--')}"
+                commit_path = model_cache_dir / model_name / "refs" / "main"
+                if not commit_path.exists():
+                    raise ValueError(
+                        "Huggingface cache not found. Please specify 'hf_cache_path' in 'enable_cloud_cache' config."
+                    )
+                with open(commit_path) as file:
+                    hf_hub_model_commit_id = file.read().strip()
+        else:
+            model_config_copy = deepcopy(model_config)
+            if model_config_copy.config.get("model_path") is not None:
+                model_config_copy.config["model_path"] = None
+        return hash_dict(
+            {
+                "hf_hub_model_commit_id": hf_hub_model_commit_id,
+                "model_config": model_config_copy.to_json(),
+                "pass_search_point": pass_search_point,
+                "input_model_hash": input_model_hash,
+            }
+        )
+
+    def exist_in_model_cache_map(self, output_model_hash: str):
+        model_path = self.cloud_cache_map.get(output_model_hash)
+        if not model_path:
+            logger.warning("Model cache not found in the cloud.")
+            return None
+        return model_path
+
+    def get_model_config_by_hash_key(self, output_model_hash: str):
+        model_config_blob = f"{output_model_hash}/model_config.json"
+        model_config_path = Path(self.cache_dir) / output_model_hash
+        model_config_path.mkdir(parents=True, exist_ok=True)
+        model_config_path = model_config_path / "model_config.json"
+
+        blob_client = self.container_client.get_blob_client(model_config_blob)
+        logger.info("Downloading %s to %s", model_config_blob, model_config_path)
+
+        with open(model_config_path, "wb") as download_file:
+            download_stream = blob_client.download_blob()
+            download_file.write(download_stream.readall())
+
+        # model config
+        with open(model_config_path) as file:
+            model_config_dict = json.load(file)
+            return ModelConfig.from_json(model_config_dict)
+
+    def update_model_cache_map(self, output_model_hash, cloud_model_path):
+        logger.info("Updating model cache map.")
+        if not self.local_map_path.exists():
+            self.local_map_path.touch()
+            with open(self.local_map_path, "w") as file:
+                file.write(b"{}")
+        with open(self.local_map_path, "r+b") as file:
+            self.cloud_cache_map = json.load(file)
+        self.cloud_cache_map[output_model_hash] = cloud_model_path
```

Comment:
The strategy here is seriously flawed. Let's discuss this offline.

Reply:
As I understand it, there is a single file to track the status of the cache i.e. a map of what’s in the cloud cache. This has a major flaw.
Consider two jobs – A & B, and t0 < t1 < t2 < t3
Job A starts at time t0 – downloads the current status of the cache
Job B start at time t1 - downloads the current status of the cache
Job B finishes at time t2 – updates the map, and uploads the new status of the cache
Job A finishes at time t3 – updates the map, and uploads the new status of the cache
NOTE: at t3, the record for Job B’s update is lost.

This is a simple illustration of the problem but when considering we could have large number of jobs running parallel, the scenario could be worse. A long running job will end up overwriting entries for all smaller jobs.
This can easily be fixed by eliminating the need for the map that tracks the cache status. Instead, each job creates its own record in the cloud using the hash which is already being computed. So, we treat the cloud cache as a directory rather than a file. There might be a few limitations to this strategy –
* Azure might have limitation on file names
  * Hash must be alphanumeric only
  * Length of the hash as string

To check whether a cache entry exists, we just need to compute the expected hash and look for the file itself. The file will contain the json object that was used to generate the hash.